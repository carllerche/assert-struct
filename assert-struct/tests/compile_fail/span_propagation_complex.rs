use assert_struct::assert_struct;

#[derive(Debug)]
struct ValueStream {
    data: Vec<i32>,
}

#[derive(Debug)]
enum Rows {
    Values(ValueStream),
}

#[derive(Debug)]
struct Response {
    rows: Rows,
}

#[tokio::main]
async fn main() {
    let resp = Response {
        rows: Rows::Values(ValueStream {
            data: vec![1, 2, 3],
        }),
    };

    // This should fail with error pointing to the specific operation
    // NOT to the entire assert_struct! call - replicating the original issue
    assert_struct!(resp.rows, Rows::Values(
        0.collect().await.unwrap(): [
            =~ 1
        ]
    ));
}