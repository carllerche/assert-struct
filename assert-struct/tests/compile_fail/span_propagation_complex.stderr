error[E0599]: `&ValueStream` is not an iterator
  --> tests/compile_fail/span_propagation_complex.rs:29:11
   |
4  |   struct ValueStream {
   |   ------------------ doesn't satisfy `ValueStream: Iterator`
...
28 | /     assert_struct!(resp.rows, Rows::Values(
29 | |         0.collect().await.unwrap(): [
   | |          -^^^^^^^ `&ValueStream` is not an iterator
   | |__________|
   |
   |
   = note: the following trait bounds were not satisfied:
           `&ValueStream: Iterator`
           which is required by `&mut &ValueStream: Iterator`
           `ValueStream: Iterator`
           which is required by `&mut ValueStream: Iterator`
note: the trait `Iterator` must be implemented
  --> $RUST/core/src/iter/traits/iterator.rs
   |
   | pub trait Iterator {
   | ^^^^^^^^^^^^^^^^^^
   = help: items from traits can only be used if the trait is implemented and in scope
   = note: the trait `Iterator` defines an item `collect`, but is explicitly unimplemented
